### ISSUE TITLE
🚂🔑

### ISSUE THREAD

**Issue by tiagodozedoze-web** [tiagodozedoze-web/tiagodozedoze#9]
"""
SovereignEngine 888 - Sistema de Integração TTS com Benchmarking Automatizado

Características:
- Warmup + Mediana de latência para medição precisa
- FileLock para operações concorrentes seguras
- Validação robusta de saída de áudio
- Monitoramento de VRAM em tempo real
- Cleanup seguro de recursos GPU
"""

import os
import sys
import json
import time
import shutil
import gc
import warnings
import logging
from dataclasses import dataclass, asdict
from typing import Optional, List, Any, Dict, Tuple
from pathlib import Path
from datetime import datetime
from contextlib import contextmanager
import torch
import numpy as np

# Tentativa de importação para áudio
try:
    import sounddevice as sd
    AUDIO_AVAILABLE = True
except ImportError:
    AUDIO_AVAILABLE = False
    warnings.warn("sounddevice não disponível - alertas de áudio desativados")

# FileLock para operações concorrentes
try:
    from filelock import FileLock, Timeout
    FILELOCK_AVAILABLE = True
except ImportError:
    FILELOCK_AVAILABLE = False
    warnings.warn("filelock não disponível - operações não serão thread-safe")

from huggingface_hub import snapshot_download, HfHubHTTPError
from transformers import pipeline

# Configuração de Logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    handlers=[
        logging.StreamHandler(sys.stdout),
        logging.FileHandler('sovereign_engine.log')
    ]
)
logger = logging.getLogger("SovereignEngine_888")

@dataclass
class BenchmarkResult:
    """Resultado estruturado de benchmark TTS"""
    latency_ms: float
    vram_peak_gb: float
    vram_avg_gb: float
    audio_duration_sec: float
    rtf: float  # Real-time factor
    timestamp: str
    status: str

class SovereignEngine:
    """
    Motor de integração e benchmarking de modelos TTS.
    
    Args:
        registry_path: Caminho para o arquivo JSON de registro
        latency_threshold_ms: Limite máximo de latência aceitável (padrão: 888ms)
        benchmark_runs: Número de runs para cálculo da mediana (padrão: 3)
    """
    
    RECOMMENDED_MODELS = {
        "speecht5": "microsoft/speecht5_tts",
        "fastspeech2": "facebook/fastspeech2-en-ljspeech",
        "bark_small": "suno/bark-small",
        "bark": "suno/bark"
    }
    
    def __init__(
        self, 
        registry_path: str = "sovereign_registry.json",
        latency_threshold_ms: float = 888.0,
        benchmark_runs: int = 3
    ):
        self.registry_path = registry_path
        self.lock_path = registry_path + ".lock"
        self.log_path = "sovereign_performance.log"
        self.modules_dir = Path("./modules")
        self.modules_dir.mkdir(exist_ok=True)
        
        self.latency_threshold_ms = latency_threshold_ms
        self.benchmark_runs = max(1, benchmark_runs)
        
        # 1. Hardware Detection PRIMEIRO
        self.device = "cuda" if torch.cuda.is_available() else "cpu"
        self.device_id = 0 if self.device == "cuda" else -1
        self.total_vram = self._get_total_vram()
        
        # 2. Inicialização de componentes
        self.tts_pipeline: Optional[Any] = None
        self.current_model_name: Optional[str] = None
        self.registry = self._load_registry()
        
        logger.info(f"SovereignEngine inicializado: device={self.device}, vram={self.total_vram:.2f}GB")

    def _get_total_vram(self) -> float:
        """Retorna VRAM total em GB"""
        if torch.cuda.is_available():
            return torch.cuda.get_device_properties(0).total_memory / (1024**3)
        return 0.0

    def _get_current_vram(self) -> float:
        """Retorna VRAM atualmente alocado em GB"""
        if torch.cuda.is_available():
            return torch.cuda.memory_allocated() / (1024**3)
        return 0.0

    def _load_registry(self) -> dict:
        """Carrega registro com FileLock se disponível"""
        if not os.path.exists(self.registry_path):
            return {
                "installed_modules": {},
                "blacklist": [],
                "last_scan": None,
                "system_info": {
                    "device": self.device,
                    "total_vram_gb": self.total_vram,
                    "initialized_at": time.strftime("%Y-%m-%d %H:%M:%S")
                }
            }
        
        try:
            if FILELOCK_AVAILABLE:
                with FileLock(self.lock_path, timeout=5):
                    with open(self.registry_path, "r") as f:
                        return json.load(f)
            else:
                with open(self.registry_path, "r") as f:
                    return json.load(f)
        except (json.JSONDecodeError, Timeout) as e:
            logger.error(f"Erro ao carregar registro: {e}. Criando novo.")
            return {
                "installed_modules": {},
                "blacklist": [],
                "last_scan": None,
                "system_info": {"device": self.device, "total_vram_gb": self.total_vram}
            }

    def _save_registry(self):
        """Salva registro com FileLock para thread-safety"""
        try:
            if FILELOCK_AVAILABLE:
                with FileLock(self.lock_path, timeout=10):
                    with open(self.registry_path, "w") as f:
                        json.dump(self.registry, f, indent=2)
            else:
                with open(self.registry_path, "w") as f:
                    json.dump(self.registry, f, indent=2)
        except Timeout:
            logger.error("Timeout ao adquirir lock para salvar registro")
            raise

    def _log_metrics(self, model_name: str, result: BenchmarkResult):
        """Log estruturado de métricas"""
        log_entry = {
            "model": model_name,
            "metrics": asdict(result),
            "system": {
                "device": self.device,
                "vram_total_gb": self.total_vram,
                "vram_current_gb": self._get_current_vram()
            }
        }
        
        with open(self.log_path, "a") as f:
            f.write(json.dumps(log_entry) + "\n")

    def cleanup_blacklist(self, max_age_days: int = 7):
        """Remove entradas antigas da blacklist com parsing seguro de datas"""
        cutoff = time.time() - (max_age_days * 86400)
        original_count = len(self.registry.get("blacklist", []))
        
        def entry_is_recent(entry: dict) -> bool:
            ts_str = entry.get("timestamp")
            if not ts_str:
                return False
            try:
                ts = datetime.strptime(ts_str, "%Y-%m-%d %H:%M:%S").timestamp()
                return ts > cutoff
            except ValueError:
                logger.warning(f"Timestamp inválido removido: {ts_str}")
                return False
        
        self.registry["blacklist"] = [e for e in self.registry.get("blacklist", []) if entry_is_recent(e)]
        removed = original_count - len(self.registry["blacklist"])
        
        if removed > 0:
            logger.info(f"Blacklist limpa: {removed} entradas removidas")
            self._save_registry()

    def _benchmark_inference(self, text: str = "Protocolo 888 ativo.") -> BenchmarkResult:
        """
        Executa benchmark completo com warmup e múltiplas runs.
        
        Returns:
            BenchmarkResult com latência mediana, VRAM e RTF
        """
        if not self.tts_pipeline:
            raise RuntimeError("Pipeline TTS não carregado")
        
        # Reset VRAM stats
        if torch.cuda.is_available():
            torch.cuda.reset_peak_memory_stats()
            vram_before = torch.cuda.memory_allocated()
        
        # Warmup (descarta primeira inferência - compilação CUDA)
        logger.debug("Executando warmup...")
        _ = self.tts_pipeline(text)
        
        if torch.cuda.is_available():
            torch.cuda.synchronize()
        
        # Benchmark runs
        latencies = []
        audio_duration = 0.0
        
        for i in range(self.benchmark_runs):
            if torch.cuda.is_available():
                torch.cuda.synchronize()
            
            start = time.perf_counter()
            result = self.tts_pipeline(text)
            
            if torch.cuda.is_available():
                torch.cuda.synchronize()
            
            latency = (time.perf_counter() - start) * 1000
            latencies.append(latency)
            
            # Calcula duração do áudio na última run
            if i == self.benchmark_runs - 1 and isinstance(result, dict):
                audio_array = np.array(result.get("audio", []))
                sr = result.get("sampling_rate", 16000)
                audio_duration = len(audio_array) / sr if len(audio_array) > 0 else 0
        
        # Estatísticas
        latency_median = float(np.median(latencies))
        latency_std = float(np.std(latencies))
        
        # VRAM stats
        vram_peak = 0.0
        vram_avg = 0.0
        if torch.cuda.is_available():
            vram_peak = torch.cuda.max_memory_allocated() / (1024**3)
            vram_avg = (torch.cuda.memory_allocated() + vram_before) / 2 / (1024**3)
            torch.cuda.reset_peak_memory_stats()
        
        # Real-time factor (RTF) - < 1.0 é tempo real
        rtf = (latency_median / 1000) / audio_duration if audio_duration > 0 else float('inf')
        
        logger.info(f"Benchmark: {latency_median:.2f}ms (±{latency_std:.2f}), "
                   f"VRAM: {vram_peak:.2f}GB, RTF: {rtf:.3f}")
        
        return BenchmarkResult(
            latency_ms=latency_median,
            vram_peak_gb=vram_peak,
            vram_avg_gb=vram_avg,
            audio_duration_sec=audio_duration,
            rtf=rtf,
            timestamp=time.strftime("%Y-%m-%d %H:%M:%S"),
            status="benchmarked"
        )

    def integrate_tts_model(self, repo_id: str, force: bool = False) -> bool:
        """
        Integra modelo TTS do HuggingFace com benchmark completo.
        
        Args:
            repo_id: ID do repositório HuggingFace (ex: "microsoft/speecht5_tts")
            force: Força reinstalação mesmo se já estiver no registro
            
        Returns:
            True se integrado com sucesso, False caso contrário
        """
        model_name = repo_id.replace("/", "_")
        install_path = self.modules_dir / model_name
        
        # Verifica se já está instalado
        if model_name in self.registry.get("installed_modules", {}) and not force:
            logger.info(f"Modelo {model_name} já instalado. Use force=True para reinstalar.")
            return True
        
        # Verifica blacklist
        blacklisted = [e for e in self.registry.get("blacklist", []) if e.get("model") == model_name]
        if blacklisted and not force:
            logger.warning(f"Modelo {model_name} está na blacklist. Use force=True para tentar novamente.")
            return False
        
        try:
            # Download
            logger.info(f"Baixando {repo_id}...")
            snapshot_download(repo_id=repo_id, local_dir=str(install_path), resume_download=True)
            
            # Carrega pipeline
            logger.info(f"Carregando pipeline TTS...")
            self.tts_pipeline = pipeline(
                "text-to-speech",
                model=str(install_path),
                device=self.device_id
            )
            self.current_model_name = model_name
            
            # Benchmark
            logger.info("Executando benchmark...")
            result = self._benchmark_inference()
            
            # Validação de latência
            if result.latency_ms <= self.latency_threshold_ms:
                # Sucesso
                self.registry["installed_modules"][model_name] = {
                    "repo_id": repo_id,
                    "latency_ms": result.latency_ms,
                    "vram_peak_gb": result.vram_peak_gb,
                    "rtf": result.rtf,
                    "path": str(install_path),
                    "integrated_at": result.timestamp
                }
                self._save_registry()
                self._log_metrics(model_name, result)
